packages/subproviders/src/subproviders/nonce_tracker.ts

```diff
@@ -0,0 +1,107 @@
+import * as _ from 'lodash';
+
+import EthereumTx = require('ethereumjs-tx');
+import ethUtil = require('ethereumjs-util');
+import providerEngineUtils = require('web3-provider-engine/util/rpc-cache-utils');
+
+import { JSONRPCPayload } from '../types';
+
+import { Subprovider } from './subprovider';
+
+const NONCE_TOO_LOW_ERROR_MESSAGE = 'Transaction nonce is too low';
+
+export class NonceTrackerSubprovider extends Subprovider {
+    private _nonceCache: { [address: string]: string } = {};
+    private static _reconstructTransaction(payload: JSONRPCPayload): EthereumTx {
+        const raw = payload.params[0];
```

Comment:
what if params is empty?

Reply:
I do not believe this is possible as it would pass through sendTransaction validation before sendRawTransaction. This would imply an upstream provider has messed up their payload. 

Better to be safe than buggy though!